dask/array/core.py

```diff
@@ -4117,7 +4118,14 @@ def concatenate(seq, axis=0, allow_unknown_chunksizes=False):
     return Array(graph, name, chunks, meta=meta)
 
 
-def load_store_chunk(x, out, index, lock, return_stored, load_stored):
+def load_store_chunk(
+    x: MutableSequence | None,
+    out: MutableSequence,
+    index: slice,
+    lock: Any,
+    return_stored: bool,
+    load_stored: bool,
+) -> MutableSequence | None:
```

Comment:
You're explicitly stating that this function demands to mutate the input parameters.
This should be treated as a red flag.

One of the big perks of mypy is that it lets you spot parameter poisoning.
Input collections should be labelled as generic read-only whenever it is reasonable to do so.
This however means that you may need to rename internal variables, e.g.

```python
def f(x):
    x = list(x)
    x.append(123)
    return x
```
becomes
```python
def f(x: Collection) -> list:
    out = list(x)
    out.append(123)
    return out
```

Reply:
In this specific case, however, the docstring and the implementation call for an array-like, which numpy does specify; sadly there's no way to label it as immutable.

Add to dask/array/numpy_compat.py:

```python
if _numpy_120:
    from numpy.typing import ArrayLike
else:
    ArrayLike = Any
```

Reply:
The return type here is quite painful, as it forces you to check ``not None`` even when you stated ``return_stored: True``.

This is a common issue, with two canonical solutions to this:

1. just return ``Any``. 
```python
def load_store_chunk(
    x: ArrayLike,
    out: ArrayLike,
    index: slice,
    lock: Any,
    return_stored: bool,
    load_stored: bool,
) -> Any:
    [implementation]
```
3. Use overload:
```python
@overload
def load_store_chunk(
    x: ArrayLike,
    out: ArrayLike,
    index: slice,
    lock: Any,
    return_stored: Literal[True],
    load_stored: bool,
) -> ArrayLike:
    ...

@overload
def load_store_chunk(
    x: ArrayLike,
    out: ArrayLike,
    index: slice,
    lock: Any,
    return_stored: Literal[False],
    load_stored: bool,
) -> None:
    ...

def load_store_chunk(x, out, index, lock, return_stored, load_stored):
    [implementation]
```
Neither is perfect and you should decide case by case which is best - to neuter mypy checks in the user code, or to have a very long override definition here. In this case I would opt for the former.

Reply:
p.s. the implementation says that x should never be None